src/System.Private.CoreLib/shared/System/Text/Unicode/Utf16Utility.Validation.cs

```diff
@@ -0,0 +1,379 @@
+
+            int numAsciiCharsConsumedJustNow = (int)ASCIIUtility.GetIndexOfFirstNonAsciiChar(pInputBuffer, (uint)inputLength);
+            Debug.Assert(0 <= numAsciiCharsConsumedJustNow && numAsciiCharsConsumedJustNow <= inputLength);
+
+            pInputBuffer += (uint)numAsciiCharsConsumedJustNow;
+            inputLength -= numAsciiCharsConsumedJustNow;
+
+            if (inputLength == 0)
+            {
+                utf8CodeUnitCountAdjustment = 0;
+                scalarCountAdjustment = 0;
+                return pInputBuffer;
+            }
+
+            // If we got here, it means we saw some non-ASCII data, so within our
+            // vectorized code paths below we'll handle all non-surrogate UTF-16
+            // code points branchlessly. We'll only branch if we see surrogates.
+            // 
+            // We still optimistically assume the data is mostly ASCII. This means that the
+            // number of UTF-8 code units and the number of scalars almost matches the number
+            // of UTF-16 code units. As we go through the input and find non-ASCII
+            // characters, we'll keep track of these "adjustment" fixups. To get the
+            // total number of UTF-8 code units required to encode the input data, add
+            // the UTF-8 code unit count adjustment to the number of UTF-16 code units
+            // seen.  To get the total number of scalars present in the input data,
+            // add the scalar count adjustment to the number of UTF-16 code units seen.
+
+            long tempUtf8CodeUnitCountAdjustment = 0;
+            int tempScalarCountAdjustment = 0;
+
+            if (Sse2.IsSupported)
+            {
+                if (inputLength >= Vector128<ushort>.Count)
+                {
+                    Vector128<ushort> vector0080 = Vector128.Create((ushort)0x80);
+                    Vector128<ushort> vectorA800 = Vector128.Create((ushort)0xA800);
+                    Vector128<short> vector8800 = Vector128.Create(unchecked((short)0x8800));
+                    Vector128<ushort> vectorZero = Vector128<ushort>.Zero;
+
+                    do
+                    {
+                        Vector128<ushort> utf16Data = Sse2.LoadVector128((ushort*)pInputBuffer); // unaligned
+                        uint mask;
+
+                        Vector128<ushort> charIsNonAscii;
+                        if (Sse41.IsSupported)
+                        {
+                            // sets 0x0080 bit if corresponding char element is >= 0x0080
+                            charIsNonAscii = Sse41.Min(utf16Data, vector0080);
+                        }
+                        else
+                        {
+                            // sets 0x8000 bit if corresponding char element is >= 0x0080
+                            charIsNonAscii = Sse2.AndNot(vector0080, Sse2.Subtract(vectorZero, Sse2.ShiftRightLogical(utf16Data, 7)));
```

Comment:
The comment here says 0x8000, the comment in SSE4.1 is 0x0080.  Are they actually different?  If so, and all that matters is non-zero, or "some high bit" then a clarifying comment would be good.  (If not, one of the comments seemingly should be fixed :smile:)